piksi_tools/console/output_list.py

```diff
@@ -0,0 +1,295 @@
+      return True
+    else:
+      return False
+
+
+class OutputList(HasTraits):
+  """This class has methods to emulate an file-like output list of strings.
+
+  The `max_len` attribute specifies the maximum number of bytes saved by
+  the object.  `max_len` may be set to None.
+
+  The `paused` attribute is a bool; when True, text written to the
+  OutputList is saved in a separate buffer, and the display (if there is
+  one) does not update.  When `paused` returns is set to False, the data is
+  copied from the paused buffer to the main text string.
+  """
+
+  # Holds LogItems to display
+  unfiltered_list = List(LogItem)
+  # Holds LogItems while self.paused is True.
+  _paused_buffer = List(LogItem)
+  # filtered set of messages
+  filtered_list = List(LogItem)
+  # state of fiter on messages
+  log_level_filter = Enum(list(SYSLOG_LEVELS.iterkeys()))
+  # The maximum allowed length of self.text (and self._paused_buffer).
+  max_len = Trait(DEFAULT_MAX_LEN, None, Int)
+
+  # When True, the 'write' or 'write_level' methods append to self._paused_buffer
+  # When the value changes from True to False, self._paused_buffer is copied
+  # back to self.unfiltered_list.
+  paused = Bool(False)
+
+
+
+  def write(self, s):
+    """
+    Write to the lists OutputList as STDOUT or STDERR.
+    This method exist to allow STDERR and STDOUT to be redirected into this
+    display. It should only be called when writing to STDOUT and STDERR.
+    Any log levels from this method will be LOG_LEVEL_CONSOLE
+    Ignores spaces.
+
+    Parameters
+    ----------
+    s : str
+      string to cast as LogItem and write to tables
+    """
+
+    if not s.isspace():
+      log = LogItem(s, CONSOLE_LOG_LEVEL)
+      if self.paused:
+        self.append_truncate(self._paused_buffer, log)
+      else:
+        self.append_truncate(self.unfiltered_list, log)
+        if log.matches_log_level_filter(self.log_level_filter):
+          self.append_truncate(self.filtered_list, log)
+
+  def write_level(self, s, level):
+    """
+    Write to the lists in OutputList from device or user space.
+
+    Parameters
+    ----------
+    s : str
+      string to cast as LogItem and write to tables
+    level : int
+      Integer log level to use when creating log item.
+    """
+    log = LogItem(s, level)
+    if self.paused:
+      self.append_truncate(self._paused_buffer, log)
+    else:
+      self.append_truncate(self.unfiltered_list, log)
+      if log.matches_log_level_filter(self.log_level_filter):
+        self.append_truncate(self.filtered_list, log)
+
+  def append_truncate(self, buffer, s):
+    """
+    Append to a front of buffer, keeping overall size less than max_len
+
+    Parameters
+    ----------
+    s : List
+      Buffer to append
+    s : LogItem
+      Log Item to add
+    """
+    if len(buffer) > self.max_len:
+      assert (len(buffer) - self.max_len) == 1, "Output list buffer is too long"
+      buffer.pop()
+    buffer.insert(0, s)
```

Comment:
So that the newest message was at the top.  The table gui element I used has no ability to sort the list while displaying.   If I use the table element that has built in sorting, each row was too large to be useful.

Also I don't think that the insert is any worse than appending in python.  I looked it up and the insert is nearly the same mutation as appending.